### ISSUE TITLE
delta time runs crazy when toggling fullscreen

### ISSUE THREAD

**Issue by fossegutten** [17cupsofcoffee/tetra#224]
**Summary:**

Assigned a hotkey to toggle fullscreen fast while debugging my project. If I do this at the same time as moving my player/camera, it suddenly moves a very long distance (300-400 pixels instantly, with a speed that should be 100 px/sec). This happens when multiplying speed with get_delta_time(ctx), right after toggling fullscreen. Did some quick debugging and I got delta times to above 0.4 seconds (probably several in a row, for this to happen).

This obviously causes serious gameplay bugs.

**Additional info:**
Windows 10

**Comment by 17cupsofcoffee** [comment]
If I remember right, Windows blocks the thread when you resize the window or something like that, so it could be that there just hasn't been a frame rendered for that long? I'll investigate further after work - if you have any code that can replicate the issue that'd be handy.

Also, out of interest, are you using `Timestep::Fixed` (the default) or `Timestep::Variable`? If it's the former, you should not use `get_delta_time` in `update`, for the reasons I detailed in [this comment](https://github.com/17cupsofcoffee/tetra/issues/211#issuecomment-732204759).

**Comment by fossegutten** [comment]
I used the default. The problem is that the player instantly moved **a lot** longer than it would do in the same timeframe if I didnt resize the screen.
Alright got it, so I should just multiply by 1.0 / 60.0 instead of delta time then.

**Comment by 17cupsofcoffee** [comment]
Yes :+1: Let me know if that resolves your issue.

Alternatively, you could switch to `Timestep::Variable`, which works more like you were expecting (`update` and `draw` always called once per frame, whatever the framerate). In that case, multiplying by `get_delta_time` *would* be the correct approach. This isn't the default because it makes your game logic [non-deterministic](https://gameprogrammingpatterns.com/game-loop.html#one-small-step,-one-giant-step) (i.e. it may output slightly different results depending on the delta time, due to floating point rounding error).

The main downside to using the fixed timestep is that it can look less smooth at high framerates than variable timestep (since you're only updating the position of things 60 times a second, rather than at the player's framerate). This is usually fixed via [interpolation or extrapolation](https://github.com/17cupsofcoffee/tetra/blob/main/examples/interpolation.rs), but that can be a bit of a pain to implement. Alternatively, you can turn up the update rate (e.g. `TimeStep::Fixed(120.0)`), but that increases the amount of processing your game has to do. 

So that's the main trade-off - variable timestep makes it easier to get smooth visuals, at the cost of making your logic non-deterministic, whereas fixed timestep is a bit more effort, but guarentees your game logic will run consistently on every PC.

I think all this needs to be better documented in the `time` module - it's mentioned on the docs for `Timestep` and `get_delta_time`, but if you don't go out of your way to click through to their doc pages, it's very easy to miss.

**Comment by fossegutten** [comment]
Thanks for explanations!
Currently i prefer the Variable mode, as it feels a bit smoother.
I added this code in update function, to be able to test both:
```
let delta = match time::get_timestep(ctx) {
            Timestep::Fixed(fps) => (1.0 / fps) as f32,
            Timestep::Variable => get_delta_time(ctx).as_secs_f32()
};
```
The original issue also happened on Ubuntu 20.04. The code above fixed it :)
